libft/sources/strings/ft_split_buf.c

```diff
@@ -0,0 +1,107 @@
+/* ************************************************************************** */
+/*                                                                            */
+/*                                                        :::      ::::::::   */
+/*   ft_split_buf.c                                     :+:      :+:    :+:   */
+/*                                                    +:+ +:+         +:+     */
+/*   By: taretiuk <taretiuk@student.42.fr>          +#+  +:+       +#+        */
+/*                                                +#+#+#+#+#+   +#+           */
+/*   Created: 2024/09/23 14:08:20 by taretiuk          #+#    #+#             */
+/*   Updated: 2024/10/04 16:42:25 by taretiuk         ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "../../sources/strings/strings_internal.h"
+
+static int	check_edges(char **res, int *cwi)
+{
+	if (!res[*cwi])
+	{
+		while (-- (*cwi) > -1)
+			free(res[*cwi]);
+		free(res);
+		return (1);
+	}
+	if (!res[*cwi][0])
+	{
+		free(res[*cwi]);
+		(*cwi)--;
+	}
+	return (0);
+}
+
+static void	calc_tokens(const char *pp, int *token_q, t_delims *ar)
+{
+	size_t		op_len;
+	int			last_op;
+	const char	*p;
+
+	*token_q = 0;
+	last_op = 0;
+	p = pp;
+	while (*p)
+	{
+		while (*p == ' ')
+			p++;
+		if (find_delimiter(p, ar, &op_len))
+		{
+			(*token_q)++;
+			p += op_len;
+			last_op = 1;
+		}
+		else
+		{
+			while (*p && !find_delimiter(p, ar, &op_len))
+				p++;
+			(*token_q)++;
+		}
+	}
+}
+
+static int	extract_token(char **res, const char **s, int token_q, t_delims *ar)
+{
+	int			cwi;
+	size_t		op_len;
+
+	cwi = 0;
+	op_len = 0;
+	while (*s && cwi < token_q)
+	{
+		while (**s == ' ')
+			(*s)++;
+		if (find_delimiter(*s, ar, &op_len))
+		{
+			extract_delimiter(res, s, op_len, cwi);
+		}
+		else
+		{
+			extract_plain_token(res, s, ar, cwi);
+		}
+		if (check_edges(res, &cwi))
+			return (1);
+		cwi++;
+	}
+	return (0);
+}
+
+char	**ft_split_buf(const char *s, void *delim_ar, int *sz)
+{
+	char		**res;
+	int			token_q;
+	t_delims	*ar;
```

Comment:
Nope. This doesn't make sense. If the user of your function wants to call this function, he'd have to create an array of t_delims. This means that this structure should be not internal, its definition should be accessible from strings.h.

Reply:
@ioannnikulin please check if my solution to include internal header in strings.h make sense? then i include trings.h in all files.

Reply:
no, it breaks the idea of internal headers. see the comment in the code.